soc: nordic_nrf: validate base addresses with new DT API

mbolivar-nordic opened the pull request:
Use the new devicetree API in a file which cross-checks dt reg base
addresses with values from the vendor HAL.

mbolivar-nordic committed:
soc: nordic_nrf: validate base addresses with new DT API

Use the new devicetree API in a file which cross-checks dt reg base
addresses with values from the vendor HAL.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
I tested it like this:

```
sanitycheck -T samples/hello_world/ \
	   -p nrf51_pca10028 \
	   -p nrf52840_pca10056 \
	   -p nrf9160_pca10090 \
	   -p nrf9160_pca10090ns \
	   -p nrf5340_dk_nrf5340_cpuapp \
	   -p nrf5340_dk_nrf5340_cpuappns \
	   -p nrf5340_dk_nrf5340_cpunet
```

Hopefully that isn't missing anything.

ioannisg commented:
> I tested it like this:
> 
> ```
> sanitycheck -T samples/hello_world/ \
> 	   -p nrf51_pca10028 \
> 	   -p nrf52840_pca10056 \
> 	   -p nrf9160_pca10090 \
> 	   -p nrf9160_pca10090ns \
> 	   -p nrf5340_dk_nrf5340_cpuapp \
> 	   -p nrf5340_dk_nrf5340_cpuappns \
> 	   -p nrf5340_dk_nrf5340_cpunet
> ```
> 
> Hopefully that isn't missing anything.

maybe you could check for the more-restricted nRF52 variants, just in case

mbolivar-nordic commented:
> maybe you could check for the more-restricted nRF52 variants, just in case

`sanitycheck -T samples/hello_world/ -p nrf52810_pca10040 -p nrf52811_pca10056 -p nrf52832_mdk` also passes

anangl commented on soc/arm/nordic_nrf/validate_base_addresses.c:
```diff
@@ -15,270 +15,101 @@
+CHECK_NODELABEL(twi2, TWI2);
+CHECK_NODELABEL(twi3, TWI3);
+CHECK_NODELABEL(uart0, UART0);
+CHECK_NODELABEL(uart1, UARTE1);
```
I'd suggest to use EasyDMA variants (i.e. SPIMx, TWIMx, UARTEx) consistently, and this way also to limit the needed translations (to ones for SPIM0-2, TWIM0-1, and UARTE0 only).

mbolivar-nordic replied:
OK, I will look at this.

anangl commented on soc/arm/nordic_nrf/validate_base_addresses.c:
```diff
@@ -15,270 +15,101 @@
+CHECK_NODELABEL(uart1, UARTE1);
+CHECK_NODELABEL(uart2, UARTE2);
+CHECK_NODELABEL(uart3, UARTE3);
+CHECK_NODELABEL(wdt, WDT0);
```
I think we should get rid of this entry and rename `wdt` labels to `wdt0` instead.

mbolivar-nordic replied:
> I think we should get rid of this entry and rename `wdt` labels to `wdt0` instead.

On second thought, I would like to skip this step for now and defer it to potential future improvements to this file.

mbolivar-nordic commented on soc/arm/nordic_nrf/validate_base_addresses.c:
```diff
@@ -15,270 +15,101 @@
+ *   compatible, with special characters converted to underscores
+ * - 'peripheral' specifies the address by NRF_<peripheral>_BASE
+ */
+#define CHECK_COMPAT(compat, mdk)					\
```
I was just trying to keep the new version working similarly to the previous one. I am happy to make this change.

mbolivar-nordic commented on soc/arm/nordic_nrf/validate_base_addresses.c:
```diff
@@ -15,270 +15,101 @@
+CHECK_NODELABEL(twi0, TWI0);
+CHECK_NODELABEL(twi1, TWI1);
+CHECK_NODELABEL(twi2, TWI2);
+CHECK_NODELABEL(twi3, TWI3);
```
Now I'm worried that sanitycheck didn't catch these. I imagine it's because the nodes aren't enabled by default on any of those boards in a hello world configuration. I will try to enable all the instances with an overlay and retest.

mbolivar-nordic committed:
dts: nordic_nrf: add flash_controller labels

Make sure every flash controller has a node label "flash_controller".
This will make it easier to refer to the SoC NVMC node when necessary.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@anangl please take another look. I have tested it on the following boards:

- nrf51dk_nrf51422
- nrf52810_pca10040
- nrf52840dk_nrf52811
- nrf52_pca10040
- nrf52840dk_nrf52840
- nrf5340_dk_nrf5340_cpuapp
- nrf5340_dk_nrf5340_cpuappns
- nrf5340_dk_nrf5340_cpunet
- nrf9160_pca10090
- nrf9160_pca10090ns

There are a couple of Python scripts in the commit log for the top commit which explain the details of the testing strategy.

mbolivar-nordic commented:
Trying to make the commit log pass gitlint.

mbolivar-nordic commented on soc/arm/nordic_nrf/validate_base_addresses.c:
```diff
@@ -1,284 +1,128 @@
-
-#if defined(DT_NORDIC_NRF_WATCHDOG_WDT_0_BASE_ADDRESS)
-CHECK_ADDRESS(DT_NORDIC_NRF_WATCHDOG_WDT_0_BASE_ADDRESS, NRF_WDT0);
+#if defined(CONFIG_SOC_NRF9160) && !defined(NRF_PDM0) && defined(NRF_PDM)
```
OK. I'd like to add a special check for nRF52 then here as well, to try to keep these blocks self-documenting for which SoCs they are required for.

mbolivar-nordic commented on soc/arm/nordic_nrf/validate_base_addresses.c:
```diff
@@ -1,284 +1,128 @@
+CHECK_DT_REG(uicr, NRF_UICR);
+CHECK_DT_REG(usbd, NRF_USBD);
+CHECK_DT_REG(vmc, NRF_VMC);
+CHECK_DT_REG(wdt0, NRF_WDT0);
```
Sorry about that. I will just remove the zeroes so we are validating what we have in place now.

anangl replied:
Sorry, I was imprecise in my comment. We need translation `#define NRF_WDT0 NRF_WDT` for this to work. And currently we use `wdt` node labels, so the check should be `CHECK_DT_REG(wdt, NRF_WDT0)`, until we update labels of watchdog nodes.

mbolivar-nordic commented on soc/arm/nordic_nrf/validate_base_addresses.c:
```diff
@@ -1,284 +1,128 @@
+CHECK_DT_REG(gpio0, NRF_P0);
+CHECK_DT_REG(gpio1, NRF_P1);
+CHECK_DT_REG(gpiote, NRF_GPIOTE);
+CHECK_DT_REG(i2s0, NRF_I2S);
```
> It should be `NRF_I2S0` and corresponding translation should be added.

Is there a rule that determines when these should be added or not? E.g. there's no translation for NRF_CLOCK0 or NRF_QDEC0. Why is I2S different?

anangl replied:
That's because for I2S we get (from MDK) NRF_I2S0 symbol defined for nRF5340 and NRF_I2S for others. We need to handle both in the same driver (nrfx_i2s.c), so we translate NRF_I2S to NRF_I2S0 in nrf_i2s.h. Same story with NRF_PDM[0] nad NRF_WDT[0]. NRF_CLOCK and NRF_QDEC are always defined without indexes.

mbolivar-nordic replied:
Ah, I missed the 5340 case. Thanks.

mbolivar-nordic commented on soc/arm/nordic_nrf/validate_base_addresses.c:
```diff
@@ -1,284 +1,134 @@
 
-#if defined(DT_NORDIC_NRF_WATCHDOG_WDT_0_BASE_ADDRESS)
-CHECK_ADDRESS(DT_NORDIC_NRF_WATCHDOG_WDT_0_BASE_ADDRESS, NRF_WDT0);
+#if !defined(NRF_WDT0) && defined(NRF_WDT)
```
Added the translation here. Gave up on tracking which SoCs things apply to.
